test/getselector.spec.ts

```diff
@@ -0,0 +1,148 @@
+    _client: puppeteer.CDPSession;
+  }
+}
+
+let browser: puppeteer.Browser,
+  page: puppeteer.Page,
+  client: puppeteer.CDPSession;
+
+describe('DOM', () => {
+  beforeAll(async () => {
+    browser = await puppeteer.launch({ defaultViewport: null, headless: true });
+    page = await browser.newPage();
+    client = page._client;
+  });
+
+  afterAll(async () => {
+    await browser.close();
+  });
+
+  describe('isSubmitButton', () => {
+    it('should return true if the button is a submit button', async () => {
+      await page.setContent(`<form><button id="button"</button></form>`);
+
+      const element = await page.$('button');
+      const isSubmitCheck = await isSubmitButton(
+        client,
+        element._remoteObject.objectId
+      );
+      expect(isSubmitCheck).toBe(true);
+    });
+
+    it('should return false if the button is not a submit button', async () => {
+      await page.setContent(`<button id="button"></button>`);
+      const element = await page.$('button');
+      const isSubmitCheck = await isSubmitButton(
+        client,
+        element._remoteObject.objectId
+      );
+      expect(isSubmitCheck).toBe(false);
+    });
+  });
+
+  describe('getSelector', () => {
+    it('should return the aria name if it is available', async () => {
+      await page.setContent(
+        `<form><button id="button">Hello World</button></form>`
+      );
+
+      const element = await page.$('button');
+      const selector = await getSelector(
+        client,
+        element._remoteObject.objectId
+      );
+      expect(selector).toBe('aria/Hello World[role="button"]');
+    });
+
+    it('should return an aria name selector for the closest link or button', async () => {
+      await page.setContent(
+        `<form><button><span id="button">Hello World</span></button></form>`
+      );
+
+      const element = await page.$('button');
+      const selector = await getSelector(
+        client,
+        element._remoteObject.objectId
+      );
+      expect(selector).toBe('aria/Hello World[role="button"]');
+    });
+
+    it.skip('should return an aria name selector for the closest link or button if the text is not an exact match', async () => {
+      await page.setContent(
+        `<form><button><span id="button">Hello</span> World</button></form>`
+      );
+
+      const element = await page.$('#button');
+      const selector = await getSelector(
+        client,
+        element._remoteObject.objectId
+      );
+      expect(selector).toBe('aria/Hello World[role="button"]');
+    });
+
+    it('should return css selector if the element is not identifiable by an aria selector', async () => {
+      await page.setContent(
+        `<form><div><span id="button">Hello</span> World</div></form>`
+      );
+
+      const element = await page.$('#button');
+      const selector = await getSelector(
+        client,
+        element._remoteObject.objectId
+      );
+      expect(selector).toBe('#button');
```

Comment:
What about a test where there'd no ID?

Reply:
I have now added a test where the button has no ID. It also shows that there's room for improvement in our CSS selector generation heuristics -- we generate the full path `body > form > div > span` although there is only this `span` tag on the page so that should be sufficient.